### PR TITLE
Migrate Tribes page to React (finishing the feature)

### PR DESCRIPTION
**Pull request by mrkvon**
#### Proposed Changes

- Finalize the migration of `/tribes` page to React
- api calls are made from React
- there is a remaining layer of angularjs for communicating between angular and react

#### Testing Instructions

- git checkout or switch to `tribes-react-page`, `npm install`, `npm start`
- go to `/tribes`
- See that the page works as expected; that the functionality has not degraded.
  - specifically, test when user is both `signed in` and `signed out`.
- FYI we have not migrated some google analytics calls. So they were removed.

Finishing work on #1133
~Depends on #1137, #1143 and #1144, and whatever these depend on. The dependencies should be merged to `tribes-react` branch first and then this branch should be rebased.~ (update: dependencies merged)

### CODE REVIEW

File: modules/tribes/client/components/TribesPage.component.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+    onMembershipUpdated(data);
+  };
+
+  useEffect(() => {
+    (async () => {
+      const tribes = await api.read();
+      setTribes(tribes);
+    })();
+  }, []);
```
[the remark this answers is not in the document]
Comment (mrkvon) [reply modules/tribes/client/components/TribesPage.component.js]: If I need to choose, I like the 3rd option of defining `fetchData` outside, or the original one. But this is the kind of stuff that's impossible to argue about. :slightly_smiling_face: 

The `useAsyncEffect` looks magical, but has some additional benefits, doesn't it? Cancellation... does it stop the async function when leaving the component? Interesting!

File: modules/tribes/client/controllers/tribes-list.client.controller.js [hunk 2]
```diff
@@ -4,39 +4,26 @@
     .controller('TribesListController', TribesListController);
 
   /* @ngInject */
-  function TribesListController(tribes, $state, Authentication, TribeService, $rootScope, $scope) {
+  function TribesListController(Authentication, $rootScope, $scope) {
```
[the remark this answers is not in the document]
Comment (mrkvon) [reply modules/tribes/client/controllers/tribes-list.client.controller.js]: Whoa! I think I may want to leave this one for a followup. But I'll add a TODO with a link to your comment here.
Or maybe I'll give it a try. :slightly_smiling_face:

File: modules/tribes/client/components/TribesPage.component.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (nicksellen) [reply modules/tribes/client/components/TribesPage.component.js]: No need to change anything at all here.

The `useAsyncEffect` isn't magical, just wrapping up the pattern you can do without it:

```js
useEffect(() => {
  let cancelled = false;
  (async () => {
    const tribes = await api.read();
    if (cancelled) return;
    setTribes(tribes);
  })();
  return () => { cancelled = true; };
}, []);
```

It doesn't actually stop the async function as they don't have that feature (there was a cancellable promises proposal, but that [got cancelled](https://github.com/tc39/proposal-cancelable-promises)), axios requests [can be cancelled](https://github.com/axios/axios#cancellation) so it's possible to connect that up too.

File: modules/tribes/client/controllers/tribes-list.client.controller.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (nicksellen) [reply modules/tribes/client/controllers/tribes-list.client.controller.js]: Yes, definitely leave it unless you particularly feel like it...

File: modules/tribes/client/components/TribesPage.component.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrkvon) [reply modules/tribes/client/components/TribesPage.component.js]: I see. The `useAsyncEffect` makes sure the component doesn't get updated after it's unmounted. If I observe correctly.

File: modules/tribes/client/components/TribesPage.component.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+export default function TribesPage({ user, onDisplayPhoto, onHidePhoto, onMembershipUpdated }) {
+  const [tribes, setTribes] = useState([]);
+
+  const handleMembershipUpdated = data => {
+    // update the tribes in state
+
+    setTribes(tribes => {
+      const i = tribes.findIndex(tribe => tribe._id === data.tribe._id);
+      return [...tribes.slice(0, i), data.tribe, ...tribes.slice(i + 1)];
```
[the remark this answers is not in the document]
Comment (mrkvon) [reply modules/tribes/client/components/TribesPage.component.js]: I like this simplification. Done in 21d2106cf30cb640d6297226aa786991b4709474

File: modules/tribes/client/views/tribes-list.client.view.html [hunk 4]
```diff
@@ -1,21 +1,6 @@
...
-        onMembershipUpdated="tribesList.broadcastChange"
-      ></tribes-list>
-    </div>
-  </div>
-
-  <tribes-join-trustroots ng-if="!app.user">
-
-</section>
+  onMembershipUpdated="tribesList.broadcastUpdatedUser"
+></tribes-page>
```
[the remark this answers is not in the document]
Comment (mrkvon) [reply modules/tribes/client/views/tribes-list.client.view.html]: Done in 21d2106cf30cb640d6297226aa786991b4709474

File: modules/tribes/client/controllers/tribes-list.client.controller.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mrkvon) [reply modules/tribes/client/controllers/tribes-list.client.controller.js]: Added a TODO in 21d2106cf30cb640d6297226aa786991b4709474
Leaving this as it is for now.

File: modules/tribes/client/components/TribesPage.component.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrkvon) [reply modules/tribes/client/components/TribesPage.component.js]: used the suggested version 3 (`fetchData` defined outside `useEffect`)
21d2106cf30cb640d6297226aa786991b4709474

The `useAsyncEffect` may have additional benefits. To be used here and elsewhere. In other PR.

**Comment by mrkvon** [comment]
@nicksellen @simison Nick's feedback is incorporated. If you like, make another round of reviews. Otherwise I'll squash and merge to tribes-react in 24+ hours. Then I'll wait for approval of the tribes-react #1148 and rebase + merge that to master. The tests can be merged in a follow-up I believe.

File: modules/tribes/client/components/TribesPage.component.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+import * as api from '../api/tribes.api';
+
+export default function TribesPage({ user, onDisplayPhoto, onHidePhoto, onMembershipUpdated }) {
+  const [tribes, setTribes] = useState([]);
+
+  const handleMembershipUpdated = data => {
+    // update the tribes in state
+
+    setTribes(tribes => tribes.map(tribe => tribe._id === data.tribe._id ? data.tribe : tribe));
```
Comment (simison) [inline modules/tribes/client/components/TribesPage.component.js]: I found this little hard to understand (a couple of arrow functions and ternary in one line, with lots of `tribe` or `tribes` sounding things going on.). It's ok to leave as is but would there be any way explaining this with comment? 😅

Not a blocker. :-)

Comment (nicksellen) [reply modules/tribes/client/components/TribesPage.component.js]: It was me that suggested to change it in this way, perhaps the clearest would be to extract a `mapReplace` function (which has come up in lodash discussion before, e.g. https://github.com/lodash/lodash/issues/1967). Something like `mapReplace(array, condition, item)`, which would replace the item at `condition` with `item`.

Comment (mrkvon) [reply modules/tribes/client/components/TribesPage.component.js]: There is also https://github.com/kolodny/immutability-helper

**Pull request merged**